crazy_rl/multi_agent/jax/base_parallel_env.py

```diff
@@ -0,0 +1,151 @@
+        state_to_dict: Translates the State into a dict.
+        step_vmap: Used to vmap step, takes the values of the state and calls step with a new State object containing
+                   the state values.
+        state: Returns a global observation (concatenation of all the agent locations and the target locations).
+
+    There are also the following functions:
+        observation_space: Returns the observation space for one agent.
+        action_space: Returns the action space for one agent.
+    """
+
+    def _observation_space(self, agent: int) -> spaces.Space:
+        """Returns the observation space of the environment. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _action_space(self, agent: int) -> spaces.Space:
+        """Returns the action space of the environment. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _compute_obs(self, state: State) -> State:
+        """Computes the current observation of the environment from a given state. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _transition_state(self, state: State, actions: jnp.ndarray, key: jnp.ndarray) -> State:
+        """Transitions the state based on the mechanics of the environment, for example makes the target move and sanitize the actions. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _sanitize_action(self, state: State, actions: jnp.ndarray) -> State:
+        """Makes the actions passed to step fit the environment, e.g. avoid making brutal moves. Must be implemented in a subclass.
+
+        Args:
+            state : the state of the environment (contains agent_location used in this function).
+            actions : 2D array containing the x, y, z action for each drone.
+        """
+        raise NotImplementedError
+
+    def _compute_reward(self, state: State) -> State:
+        """Computes the current reward value(s) from a given state. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _compute_terminated(self, state: State) -> State:
+        """Computes if the game must be stopped because the agents crashed from a given state. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _compute_truncation(self, state: State) -> State:
+        """Computes if the game must be stopped because it is too long form a given state. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def reset(self, key: jnp.ndarray) -> State:
+        """Resets the environment in initial state. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def auto_reset(self, state: State) -> State:
+        """Returns the State reinitialized if needed, else the actual State. Must be implemented in a subclass.
+
+        The values contained by State are passed in argument and used like a dictionary
```

Comment:
So the type of state is not a `State` but a `Dict` ? 

Also, move this function next to `step_vmap` and tell something like "this function handles states as dictionary, see also the pydoc of step_vmap"

Reply:
It was a mistake: it's not state but **state, with the values contained by state, which can be used like a dictionary but it's not exactly a dictionary so I don't know if this detail is important to tell in the docstring